### ISSUE TITLE
Establish uniqueness criteria for tables

### ISSUE THREAD

**Issue by robintw** [debrief/pepys-import#379]
This is really a 'point for discussion' issue with @IanMayo , with the hope that I will transfer this information to the wiki once the details have been agreed. We may also want to enforce some of these uniqueness criteria - I'll look into how to do this across multiple fields on both Postgres and SQLite.

The reason for establishing these criteria is two-fold:
 - We need to know what is valid to search by in the code. For example, it wasn't valid to search just by sensor name, as that isn't unique - you need to specify a platform too. That has been fixed, but we've got similar problems with searching for other entities such as Platform (#362)
 - When merging the databases (#375) we need a way to work out whether rows are the same or different. For example, if we have two rows from the Platform table, we need to work out whether they refer to the same platform. This requires understanding what makes a Platform unique.

So, the list below says what combination of fields make an entry unique across various tables. It's useful to read this list in conjunction with the [database schema diagram](https://docs.google.com/drawings/d/1kEEYdoU5H7rcvxCaf9Aeyh0ujvEizwZWJE4uNGGNK5o/edit).

**Reference tables:**
All reference tables are unique on the `name` field. This is already enforced by uniqueness constraints.

**Metadata tables:**
 - **Platforms**: unique on the combination of [`Name`, `Nationality`, `Type`] - that is, there are no duplicated platform names within a specific nationality and platform type (eg. not two UK submarines called Dolphin, but you can have a UK submarine and a UK frigate both called Dolphin)
 - **Sensors**: unique on the combination of [`Name`, `Host`] - that is, each Platform should have unique names for its sensors
 - **Datafiles**: unique on the combination of [`Reference`, `Hash`, `Size`]
 - **Synonyms**: unique on the combination of [`Table`, `Synonym`] - that is, each synonym must be unique for a specific table (eg. you can't have two Sensor synonyms with the same name)

**Measurement tables:**
These are harder to establish - and it partly depends on the answers in #375 as to whether we need to work out a way to establish uniqueness of this. The simplest is just to consider all fields when deciding uniqueness - presumably there should never be a case where all fields are equal for multiple rows?
 - **States**: TODO ?
 - **Contacts**: TODO ?
 - **Comments**: TODO ?

**Implementation:**
If we're going to rely on these uniqueness constraints, then we should probably enforce them in the database. We can do this with code like this in the table definition:

```
__table_args__ = (UniqueConstraint('customer_id', 'location_code', name='_customer_location_uc'))
```

See [StackOverflow](https://stackoverflow.com/questions/10059345/sqlalchemy-unique-across-multiple-columns) and [the docs](https://docs.sqlalchemy.org/en/13/core/constraints.html#sqlalchemy.schema.UniqueConstraint) for details.

**Comment by IanMayo** [comment]
## Reference tables
We do have an issue with `geometry_subtypes`. This data isn't currently supported, so its all theoretical, but I think the uniqueness across at the `parent/name` pairing.  So, two parent types may have a `bounds` sub-type.

## Metadata tables
* Sensors - yes
* datafiles - we may be missing something.  Two files are identical size, and hash, but different filenames.  I _don't think_ we should import the duplicate.  I don't think this would necessarily cause a great performance drop - we're already comparing to every previous datafile name, we'd just be doing `OR this_size == that_size` (I appreciate this may be contrary to our existing implementation, but it seems the sensible approach). 
* synonyms - we may have a problem here.  "GPS-B" may be the synonym for "GPS-2" on "HMS DRAGON".  But, "GPS-B" may be the synonym for another sensor on another platform. So, maybe there are some tables we can't support synonyms for. I'll have to give this more thought.
* platforms - the client is thinking this through.  There are complexities for ship Pennant Numbers, and maybe even more complexity for Aircraft call-signs.  I'll get back to you when I hear more.

## Measurement tables
No, we don't have to worry about duplicates here.  The duplication is handled at the datafile level.

**Comment by robintw** [comment]
Thanks @IanMayo.

I'm a little confused about the datafile uniqueness. At the moment, in the FileProcessor we use the following to decide if datafiles are unique:

```
is_loaded_before = (
            self.session.query(self.db_classes.Datafile)
            .filter(self.db_classes.Datafile.size == file_size)
            .filter(self.db_classes.Datafile.hash == file_hash)
            .first()
        )
```

That is, we're just checking the hash and the size match, and if so, we're not importing it (so if it is the same file but with a different filename, it won't be imported). Is this what you want?

I can see the problem with synonyms - I hadn't spotted that. There might be a potential solution of adding a `extra_field_name` and `extra_field_value` column to the Synonyms table. When dealing with something like a sensor, which needs to have synonyms done on a per-platform level then `extra_field_name` would be set to `host` and `extra_field_value` to the platform ID that the synonym was valid for. Then, when we search synonyms we look to see if there is an extra field specified and if so, add a filter by that field when searching. That's probably the most generic solution I can think of - what do you think? (I'll create a separate issue for this anyway, because this is actually a bug at the moment)

**Comment by IanMayo** [comment]
@robintw - in that case, the current implementation is fine.

**Comment by IanMayo** [comment]
We know this issue is blocked until we get answers from the client subject-matter-experts.

**Comment by IanMayo** [comment]
@Robin - I've just learned that we will not have a pennant number for helicopters. They are referred to by airframe number (example in [this report](https://www.royalnavy.mod.uk/news-and-latest-activity/news/2016/june/21/160621-chf-merlins-complete-move) )

So, for platform uniqueness we currently use:
* nationality
* platform name
* pennant name (if present)

**Comment by robintw** [comment]
The uniqueness criteria I'm using currently in my merge PR (#375) are:

- `Datafiles`: `size` and `hash
- `Sensors`: `name` and `host`
- `Platform`: `name`, `nationality`, `pennant`

This means when comparing entries in these tables between the master and slave database, we're only comparing on those fields - and if they match we're counting the entries as matching. All other tables are using all fields for comparison, except for measurement tables which are treated differently by comparison through datafile_ids.

Does this sound ok for the moment, until we get further information from the client? I'm going to need to know the final uniqueness criteria for platforms before I can tackle #362 properly.

**Comment by IanMayo** [comment]
Yes, that looks right.

**Comment by robintw** [comment]
At the moment, the implementation of the unique constraints is contained in the merging databases PR (#405). Is it worth extracting that code and doing a separate PR just to add those unique constraints now - or should we wait until we have definite confirmation from the client about the uniqueness for Platforms? By that point the merging databases PR might actually be ready to merged, so it might be a moot point anyway.

One thing that will need adding is code to deal with errors if we try to add duplicate entries that violate the unique constraints.

**Comment by robintw** [comment]
Just a note, that `Synonyms` also needs a unique constraint adding. I think it should be on `name` and `table` - that is, you can only have one Synonym with a specific name for a specific table. This means you can't have two Synonyms both called "Synonym1" as synonyms for the Platforms table - but you can have two Synonyms both called "Synonym1" if they're for different tables.

**Comment by robintw** [comment]
The uniqueness constraints as we currently understand them are included in the 'merging databases' PR (#405) which has now been merged, so I think this issue has been completed. If we get given updated information on uniqueness from the client then we can open a new issue to deal with that.
